pkg/provisioner/controller.go

```diff
@@ -638,3 +648,62 @@ func (c *obcController) objectBucketForClaimKey(key string) (*v1alpha1.ObjectBuc
 	}
 	return ob, nil
 }
+
+func updateSupported(old, new *v1alpha1.ObjectBucketClaim) bool {
+	// The only field supported for update is obc.spec.additionalConfig
+	if reflect.DeepEqual(new.Spec, old.Spec) {
+		return false
+	}
+	// create copy of old spec, and set the new spec's additionalConfig on it
+	oldspec := old.Spec.DeepCopy()
+	overwriteAdditionalConfig(new.Spec.AdditionalConfig, oldspec.AdditionalConfig)
+	if !reflect.DeepEqual(*oldspec, new.Spec) {
+		// new OBC spec has changed something other than additionalConfig
+		log.Error(nil, "invalid changes to OBC. only additionalConfig can be updated")
+		return false
+	}
+	return true
+}
+
+func (c *obcController) handleUpdateClaim(key string, obc *v1alpha1.ObjectBucketClaim, class *storagev1.StorageClass) error {
+	ob, err := getObjectBucketFromClaimKey(key, obc, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
+	if err != nil {
+		log.Error(err, "getting ob failed", "key", key)
+		return err
+	}
+
+	log.Info("Additional config changed, hence updating OB")
+	tmp := make(map[string]string)
+
+	tmp = ob.Spec.Endpoint.AdditionalConfigData
```

Comment:
In Go, maps need to be deep copied. This will only change the pointer, and `tmp` will therefore be overwritten by `overwriteAdditionalConfig`. If updating the object bucket fails, this will not properly reset the value.

See my test here: https://play.golang.org/p/z0BurgSZEM8

Given this potential bug, it would be best to have a unit test that verifies the provisioner is properly reset when updating the OB fails.

Reply:
Yes you are right! Sorry @thotz

Reply:
@BlaineEXE : I agree it is good to have a unit test. Sorry but I don't understand which function I need write the unit test.
Is it for `overwriteAdditionalConfig()` with set of string maps like in ur example. Or for the `handleUpdateClaim()`? 
If it is `handleUpdateClaim()`
First I need to create OBC and then call `handleUpdateClaim()` with that OBC for different test cases?

Reply:
I think for `handleUpdateClaim` and make sure the unit test coverage report activates the lines below
```go
if err != nil {
		log.Error(err, "updating OB failed, reverting provisioner to original value")
    		overwriteAdditionalConfig(tmp, ob.Spec.Endpoint.AdditionalConfigData)
    		err = c.provisioner.Update(ob)
}
```